File: lecture_7/Lecture 7. Source Code/launch_tutorial/launch_tutorial/random_velocity_node.py
import rclpy
from rclpy.node import Node
from geometry_msgs.msg import Twist
import random

class RandomVelocityNode(Node):
    def __init__(self):
        super().__init__('random_velocity_node')
        self.publisher_ = self.create_publisher(Twist, '/turtle1/cmd_vel', 10)
        self.declare_parameter('cmd_vel_rate', 3.0)
        self.timer_period = 1.0 / self.get_parameter('cmd_vel_rate').get_parameter_value().double_value
        self.timer = self.create_timer(self.timer_period, self.timer_callback)

    def timer_callback(self):
        msg = Twist()
        msg.linear.x = random.uniform(-2.0, 2.0)
        msg.angular.z = random.uniform(-2.0, 2.0)
        self.publisher_.publish(msg)
        self.get_logger().info('Publishing: "linear.x: %f, angular.z: %f"' % (msg.linear.x, msg.angular.z))

def main(args=None):
    rclpy.init(args=args)
    random_velocity_node = RandomVelocityNode()
    rclpy.spin(random_velocity_node)
    random_velocity_node.destroy_node()
    rclpy.shutdown()

if __name__ == '__main__':
    main()

File: lecture_7/src/launch_tutorial_cpp/src/random_velocity_node.cpp
#include "rclcpp/rclcpp.hpp"
#include "geometry_msgs/msg/twist.hpp"
#include <chrono>
#include <functional>
#include <random>

class RandomVelocityNode : public rclcpp::Node
{
public:
    RandomVelocityNode() : Node("random_velocity_node"), gen_(rd_()), dis_(-2.0, 2.0)
    {
        publisher_ = this->create_publisher<geometry_msgs::msg::Twist>("turtle1/cmd_vel", 10);
        this->declare_parameter("cmd_vel_rate", 3.0);
        double cmd_vel_rate = this->get_parameter("cmd_vel_rate").as_double();
        auto timer_period = std::chrono::duration<double>(1.0 / cmd_vel_rate);
        timer_ = this->create_wall_timer(timer_period, std::bind(&RandomVelocityNode::timer_callback, this));
    }
private:
    void timer_callback()
    {
        auto msg = geometry_msgs::msg::Twist();

        msg.linear.x = dis_(gen_);
        msg.angular.z = dis_(gen_);

        publisher_->publish(msg);

        RCLCPP_INFO(this->get_logger(), "Publishing: linear.x: %f, angular.z: %f", msg.linear.x, msg.angular.z);
    }
    rclcpp::Publisher<geometry_msgs::msg::Twist>::SharedPtr publisher_;
    rclcpp::TimerBase::SharedPtr timer_;

    std::random_device rd_;
    std::mt19937 gen_;
    std::uniform_real_distribution<> dis_;
};

int main(int argc, char *argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<RandomVelocityNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
}